Make Attack activities explicitly account for FrozenActors.

pchote opened the pull request:
With the dust from #13995, #15176 and all the related PRs sufficiently settled, we can now progress to fixing the actor-side logic.

The workarounds that replaced FrozenActors with Actors in the order resolving, and to allow the attack activities to target actors that are hidden under the fog have been removed, and the attack activities can now fire on FrozenActors directly.  The FrozenActor &lrarr; Actor switching when a structure becomes/loses visibility is handled by making the activity `Target`s non-readonly and introducing a `Target.Recalculate` extension method that switches between them when needed.  This also lets us propagate targets across Actor &rarr; Actor transformations for (almost) free.

Fixes #11265.
Fixes #14872.
Fixes #12856.
Fixes #10004.
Fixes #10887.
Fixes #14616.
Supersedes #12333 / #13363.

This fixes a couple of high-impact bugs in the RA mod (artillery sometimes rushing forward when ordered to attack, units refusing to target camo-pillboxes), and one of our biggest blockers for TS (stealth generators make the camo-pillbox bug affect everything).  I'd therefore like to get this in for the playtest, and once the dust settles we can migrate the remaining activities over to the new system for Next + 1.

One of the side-effects of these fixes is that artillery in the "attack everything" stance will now automatically target structures within range that are under the fog, and any attacks (including manually ordered ones) will only automatically stop once another actor reveals that the target is dead.  This has caused some controversy in IRC as it is arguably a gameplay regression.  I'm moving forward with that here because we can't think of a better alternative: never fixing the targeting isn't viable, and dropping targets when they become hidden under fog seems like a bigger gameplay regression (but does remain an option if needed).

pchote committed:
Add Actor.ReplacedByActor to track transformations.

This isn't great conceptually, but has precedent
in the Generation number.

pchote committed:
Cache FrozenActorLayer on the Player object.

This avoids unnecessary trait queries.

pchote committed:
Tweak FrozenActorLayer queries:

- FrozenActorsInRegion now filters for valid and (optionally) visible FAs
- Add new FrozenActorsInCircle to mirror World.FindActorsInCircle.

The first change means that SupportPowerDecision now correctly ignores
FrozenActors that the AI has not discovered.

pchote committed:
Add AutoTarget support for FrozenActors.

pchote commented:
Good catch. I have added a new commit which fixes that and also #14616.

The potential regression surface area has gone up, but IMO this is still more than worthwhile.

pchote commented:
Fixed. The crash was introduced in the last commit, and only happened if the explored map option was enabled.

netnazgul commented:
Tested this on two comp stomps. No ground-breaking regressions spotted yet apart from the one mentioned in the end of the OP. Which IS a gameplay-changing thing, but one I think actually makes the game better as now your units won't have the knowledge about buildings that you don't have - if you see a building under fog you are able to attack it and can continue to attack it even if it was destroyed/sold. Shift-clicking defences with V2RLs and moving on won't work now, you'll have to bring in a scout as well or else your rockets will keep falling onto the cell that once was a flame tower you'd destroyed. Also, you can't send an infantry dispatch on AssaultMove against a base scouted before, because it will get stuck on the first under-fog building as it is able to fire further than it can see.

pchote commented:
From that, I gather that we may want to rebalance the RA mod infantry to have weapon ranges that match their sight ranges before the playtest.

pchote commented:
Another option could be to add specific that can be enabled on non-artillery style units that make them move within their sight range instead of stopping at their max weapon range when in the AttackAnything stance / Assault Moving.

netnazgul commented:
Being able to select buildings under fog (but not all of them):
![openra-frozen-actors-1](https://user-images.githubusercontent.com/1614714/49328277-c84d6580-f57e-11e8-808a-2f8744a2f1f5.gif)

Artillery fires onto a spot where harvester has died before, then is sent on AssaultMove top (notice how one piece moves forward while other keep shooting the barracks even though those are destroyed):
![openra-frozen-actors-2](https://user-images.githubusercontent.com/1614714/49328312-8ffa5700-f57f-11e8-91f4-2c431fd72153.gif)

pchote commented:
@netnazgul I expect that all of these regressions will be caused by the change in 8902dfab5c3d6bc0f24e7125500d926e96fca74e. Can you please try testing on 3d2687b596e79375d1ecda44f8a5c2bb1e208fc0 ([zip link if required](https://github.com/OpenRA/OpenRA/archive/3d2687b596e79375d1ecda44f8a5c2bb1e208fc0.zip) to see if those problems go away?

matjaeck commented:
Good job @netnazgul (and sorry for my useless approving review, just don't take the stuff I write too seriously).

netnazgul commented:
I don't encounter the issues mentioned earlier anymore, so I consider that was fixed. Will conduct further testing later.

netnazgul commented:
Now units (except artillery) require either direct vision or active GPS to fire at buildings under fog. This will mean that you won't be able to attack turrets with infantry from a distance for example - they will automatically come closer to get into the vision range.

pchote commented:
You can use the force-attack modifier to do that.

netnazgul commented:
No, you can't. If you use a force-attack on a building under fog, it changes to attack (it is confirmed by the building flashing) and the same logic applies.

pchote commented:
This is something that I took pains to implement and test, and as far as I can tell it works as it should. Can you please provide some specific testcases that fail?

The weapon vs sight range on most RA infantry isn't big enough for them to be able to consistently fire into the fog from all facings, but you can bump up the weapon range to see it working.  The attack cursor will switch from the out-of-range version to the in-range version when you hold down the modifier.

teinarss commented:
> 
> 
> No, you can't. If you use a force-attack on a building under fog, it changes to attack (it is confirmed by the building flashing) and the same logic applies.

Im not experience this, just tested to attack a power plant under fog from the left side with a rocket solider in TD

teinarss commented:
I think that MSAM (Rocket Launcher) should also have TargetFrozenActors: True defined in the TD mod.

Ping @AoAGeneral

pchote commented:
Ping @OpenRA/engine-hackers and testers. If there are no further show-stoppers then it would be better to merge this so that bugs can be discovered on bleed than to have it sitting ignored on the PR queue.

netnazgul commented:
Got a crash when playing this (make clean/make all beforehand)
https://pastebin.com/SrHaMWUQ

netnazgul commented:
Also there is another potentially annoying feature that will be hard to approach: targeting units that get out of vision. 
One of two bad outcomes is happening - either your unit loses the order alltogether (when it was standing still when given order to target), or the attack order gets converted into a move order (when it was moving during retargeting). So on a periodic vision (like Yak's vision) you practically have no option to attack at all because the order will be dropped each time. You can sometimes force fire the cell when units there are out of your vision, but force-firing while you see the unit makes that a simple attack with the same problem as above. On the gif below I was unable to attack 2 rocket soldiers at all.
![openra_fog-actors-attack](https://user-images.githubusercontent.com/1614714/49750906-2683f200-fcbd-11e8-9c58-f049c3add4bc.gif)

On an unrelated note this happened. Replay viewer shows 200% speed as "194% speed" at this moment, so probably something else broke in the process. I can provide replay for this:
![openra_fog-actors-attack2](https://user-images.githubusercontent.com/1614714/49751273-086ac180-fcbe-11e8-9a49-d50622f5ee35.gif)

And in the end the game finished with a crash identical to the one above, with lots of my units in an assault-move.

pchote committed:
Track visibility modifiers on FrozenActors.

pchote commented:
> Got a crash when playing this 

Good catch.  This occurred when a unit in the ReturnFire stance was attacked by a base defense.  Now fixed.

> Also there is another potentially annoying feature that will be hard to approach: targeting units that get out of vision.

This is #10223, which isn't related to the changes here.

reaperrr reviewed (dismissed):
Code looks good as far as I'm able to tell, didn't encounter any issues

abcdefg30 commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -47,6 +47,7 @@ public class FrozenActor
 		readonly IHealth health;
 
 		public bool Visible = true;
+		public bool Hidden = false;
```
Can we get an explanation (comment) here for what this does? One would assume this is always just `!Visible`.

abcdefg30 commented on OpenRA.Game/Traits/Target.cs:
```diff
@@ -80,13 +80,20 @@ public TargetType Type
+					return frozen.IsValid && frozen.Visible && !frozen.Hidden;
+				case TargetType.Invalid:
+					return false;
+				default:
```
Can we explicitly list `TargetType.Terrain` here instead of `default` to bring this in line with what we do everywhere else in the class and to make the code easier to understand?

abcdefg30 commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -31,8 +31,8 @@ public abstract class AttackBaseInfo : PausableConditionalTraitInfo
-		[Desc("Does not care about shroud or fog. Enables the actor to launch an attack against a target even if he has no visibility of it.")]
-		public readonly bool IgnoresVisibility = false;
+		[Desc("Allow firing into the fog to target frozen actors without requiring force-fire.")]
+		public readonly bool TargetFrozenActors = false;
```
Doesn't this need an update rule?

pchote replied:
From above:
> Re lack of update rules: we can't access weapon stats from the actor updates, so theres no way to detect and warn about units that may need to have TargetFrozenActors enabled. I think the best way to handle this is with an entry in the SDK update notes, which is how we have handled similar changes in the past.

abcdefg30 replied:
I see, but I still think we want a rule that removes leftover `IgnoresVisibility` nodes.

pchote replied:
Done.

pchote commented:
Added a `RemoveAttackIgnoresVisibility` update rule as requested.

pchote committed:
Allow Attack activities to target FrozenActors directly.

Removing the legacy FrozenActor to Actor workaround
fixes a number of long-standing bugs.

This also prevents units from losing their target when
it transforms into a different actor type.

pchote commented:
Updated to split the `IgnoresVisibility` changes into their own commit. Note that the commits tab is now showing them in the wrong order.

reaperrr commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -31,8 +31,6 @@ public abstract class AttackBaseInfo : PausableConditionalTraitInfo
 
-		[Desc("Does not care about shroud or fog. Enables the actor to launch an attack against a target even if he has no visibility of it.")]
-		public readonly bool IgnoresVisibility = false;
 
```
There are 2 empty lines now.

reaperrr commented on OpenRA.Mods.D2k/Traits/AttackSwallow.cs:
```diff
@@ -67,5 +69,19 @@ public override void DoAttack(Actor self, Target target, IEnumerable<Armament> a
+			public SwallowTarget(Actor self, Target target, bool allowMovement, bool forceAttack, int facingTolerance)
+				: base(self, target, allowMovement, forceAttack, facingTolerance) { }
+
+			// Worms ignore visibiity, so don't need to recalculate targets
```
visibi*l*ity

reaperrr commented on OpenRA.Mods.D2k/Traits/AttackSwallow.cs:
```diff
@@ -67,5 +69,19 @@ public override void DoAttack(Actor self, Target target, IEnumerable<Armament> a
+				: base(self, target, allowMovement, forceAttack, facingTolerance) { }
+
+			// Worms ignore visibiity, so don't need to recalculate targets
+			protected override bool IgnoresVisibility { get { return false; } }
```
This should return `true`!

pchote replied:
Ah, good catch.  This is deleted in the next commit, so I didn't bother to test it.

pchote committed:
Remove AttackBase.IgnoresVisibility.

This was a workaround for D2K sandworms, which is
now implemented using a custom attack activity.

pchote commented:
Fixed.

obrakmann commented on OpenRA.Mods.Common/Traits/AutoTarget.cs:
```diff
@@ -269,48 +269,67 @@ public Actor ScanForTarget(Actor self, bool allowMove)
+				}
+				else if (target.Type == TargetType.FrozenActor)
+				{
+					if (attackStances == OpenRA.Traits.Stance.Enemy && self.Owner.Stances[target.FrozenActor.Owner] == OpenRA.Traits.Stance.Ally)
```
How about a `Target.AppearsHostileTo()`?

obrakmann replied:
false, I guess. But it was just an off-the-cuff suggestion, so feel free to ignore it

pchote replied:
Ok. Lets leave it as is for now, then we can reevaluate this if we find a second usecase for such a method.

obrakmann commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -31,8 +31,8 @@ public abstract class AttackBaseInfo : PausableConditionalTraitInfo
-		[Desc("Does not care about shroud or fog. Enables the actor to launch an attack against a target even if he has no visibility of it.")]
-		public readonly bool IgnoresVisibility = false;
+		[Desc("Allow firing into the fog to target frozen actors without requiring force-fire.")]
+		public readonly bool TargetFrozenActors = false;
```
Are there examples where Actors have longer firing range than vision where we would disable this? Or in other words, can we take the burden of remembering having to set this from the modder and automatically detemine it via code instead?

pchote replied:
Unfortunately yes, several.  RA infantry set a weapon range ~1 cell larger than their range, so depending on their subcell and angle of attack they can sporadically get hung-up on frozen dead buildings if assault-moved into an enemy base.

obrakmann reviewed (approved):
Let's merge it then, see what falls out.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8423039fcab126fd3aad942c7a7b28d89b8e995d)
